### PR TITLE
Add ServiceBindingAccessor that fetches service bindings from Spring's application properties

### PR DESCRIPTION
**Pull request by drichtsteiger**
## Context
Currently local tests require a "default-env.json" in addition to any configuration maintained at for example "application.yaml". This means application configuration is spread across different files. The new accessor allows to maintain the service bindings at the "application.yaml" file like

```
[...]
spring:
  config.activate.on-profile: local
cds:
  serviceBindings:
    xsuaa:
       name: xsuaa_clone
       tags: xsuaa
          credentials:
            certUrl: https://localhost:9000
            clientId: test-clientid!b123
            [...]
```

It would also be possible to make the SapVcapServicesServiceBindingAccessor obsolete as Spring Boot comes with a built-in VCAPS parser:

```
[...]
spring:
  config.activate.on-profile: cloud
cds:
  serviceBindings:
    xsuaa:
       name: ${vcap.services.xsuaa.name}
       tags: ${vcap.services.xsuaa.tags
          credentials:
            certUrl: ${vcap.services.xsuaa.credentials.certurl}
            clientId: ${vcap.services.xsuaa.credentials.clientid}
            [...]
```

### CODE REVIEW

File: bom/pom.xml [hunk 1]
```diff
@@ -74,6 +77,16 @@
...
+            <dependency>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot</artifactId>
+                <version>${spring-boot.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.springframework</groupId>
+                <artifactId>spring-core</artifactId>
+                <version>${spring.version}</version>
+            </dependency>
```
Comment (beckermarc) [inline bom/pom.xml:89]: Doesn't this come transitively with Spring Boot? I'd try to define the Spring Boot version only.

File: bom/pom.xml [hunk 2]
```diff
@@ -213,7 +239,8 @@
                             <configuration>
                                 <rules>
                                     <bannedDependencies>
-                                        <message>Do not use deprecated dependencies.</message>
+                                        <message>Do not use deprecated
+											dependencies.</message>
```
Comment (beckermarc) [inline bom/pom.xml:243]: quite a few weird formatting changes in this file..

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        public Map<String, Object> getCredentials() {
+            if (credentials == null) {
+                return Collections.emptyMap();
+            } else {
+                /* the key names are converted to lower-case */
+                return credentials
+                        .entrySet()
+                        .stream()
+                        .map(entry -> new AbstractMap.SimpleEntry<>(entry.getKey().toLowerCase(), entry.getValue()))
+                        .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:76]: Why do you normalize credentials? This might not fit with the expectation that developers have about their credential properties in the service binding. These are usually case-sensitive for all other cases (VCAP_SERVICES or Kyma bindings)

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(SapSpringPropertiesServiceBindingAccessor.class);
+
+    @Nonnull
+    @Override
+    public List<ServiceBinding> getServiceBindings()
+            throws ServiceBindingAccessException {
+        if (SapServiceBindingsPropertiesAccessor.getServiceBindingsProperties().isEmpty()) {
+            logger.info("No service bindings found in properties.");
+            return Collections.emptyList();
+        }
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:36]: I think this method (except of the log line) would do the same if this code wasn't there. Maybe simplify this a little bit by retrieving properties only in a single place in `fetchServiceBindings`

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+        List<ServiceBinding> serviceBindings = fetchServiceBindings();
+
+        serviceBindings.forEach(binding ->
+                logger.info("Service binding {} found in properties.", binding.getServiceName().get()) //NOSONAR
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:41]:
```suggestion
                logger.debug("Service binding {} found in properties.", binding.getServiceName().get()) //NOSONAR
```

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            logger.error("Service binding properties with no name detected.");
+            isValid = false;
+        }
+
+        if (serviceBindingProperties.getLabel() == null || serviceBindingProperties.getLabel().isEmpty()) {
+            logger.error("Service binding properties of {} has no label.", serviceBindingProperties.getName());
+            isValid = false;
+        }
+
+        if (serviceBindingProperties.getTags() == null || serviceBindingProperties.getTags().length == 0) {
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:92]: Why are tags mandatory? Could default to an empty tag list..

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    private boolean validateServiceBindingProperties(@Nonnull final ServiceBindingProperties serviceBindingProperties) {
+        boolean isValid = true;
+        if (serviceBindingProperties.getName() == null || serviceBindingProperties.getName().isEmpty()) {
+            logger.error("Service binding properties with no name detected.");
+            isValid = false;
+        }
+
+        if (serviceBindingProperties.getLabel() == null || serviceBindingProperties.getLabel().isEmpty()) {
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:87]: Label doesn't seem to be used in the final binding. I think in OSB `label` is the same as `service`

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+        /**
+         * name of the service
+         */
+        private String serviceName;
+
+        /**
+         * name of the instance
+         */
+        private String instanceName;
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:60]: Instane Name doesn't seem to be used anywhere

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    public static class ServiceBindingProperties {
+        /**
+         * properties for credentials
+         */
+        private Map<String, Object> credentials;
+
+        /**
+         * label of the service binding
+         */
+        private String label;
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:40]: Label doesn't seem to be (reasonably) used anywhere

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    @Nullable
+    private ServiceBinding toServiceBinding(
+            @Nonnull String serviceName,
+            @Nonnull final ServiceBindingProperties serviceBindingProperties) {
+        if (validateServiceBindingProperties(serviceBindingProperties)) {
+            return DefaultServiceBinding
+                    .builder()
+                    .copy(Collections.emptyMap())
+                    .withName(serviceBindingProperties.getName())
+                    .withServiceName(serviceName)
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:67]: I think the name should rather be the key instead of the serviceName in the properties map. There might be multiple bindings of the same service with different names. This is currently not possible.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+public class SapServiceBindingEnvironmentPostProcessor implements EnvironmentPostProcessor
+{
+    private static final Logger log = LoggerFactory.getLogger(SapServiceBindingEnvironmentPostProcessor.class);
+
+    @Override
+    public void postProcessEnvironment( ConfigurableEnvironment environment, SpringApplication application )
+    {
+        Bindable<SapServiceBindingsProperties> bindable = Bindable.of(SapServiceBindingsProperties.class);
+
+        BindResult<?> bindResult = Binder.get(environment).bind("cds", bindable);
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:26]: Please don't use the `cds` subsection, as this library is independent of CAP Java.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        if( bindResult.isBound() ) {
+            SapServiceBindingsPropertiesAccessor
+                .setServiceBindingsProperties(((SapServiceBindingsProperties) bindResult.get()).getServiceBindings());
+        } else if( bindable.getValue() != null ) {
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:30]: When would `bindable.getValue()` be set, but `bindResult.isBound()` returned false?

**Review by beckermarc — DISMISSED** [review]

File: bom/pom.xml (hunk 1 above)
Comment (drichtsteiger) [reply bom/pom.xml:89]: we have here a show-stopper: this library is compiled for Java 8, but Spring requires at Java 17.

Comment (beckermarc) [reply bom/pom.xml:89]: @Johannes-Schneider Do you see a possibility to increase the Java version of this library to 17? AFAIK CAP, Cloud SDK and XSUAA - which are the main stakeholders of this library at the moment - are all on Java 17 already.

Comment (beckermarc) [reply bom/pom.xml:89]: + @MatKuhr: Any opinion here?

Comment (MatKuhr) [reply bom/pom.xml:89]: I think we can increase to Java 17 as minimum and indicate this a bit by increasing to `0.20.0` or something like that. Since it's not a stable version and most users (if not all) are on 17 already, that should be okay. But probably better to do in a separate PR.

On a side note, I would consider to leave out the spring version in the BOM, because this library doesn't need very frequent updates, and the spring version here will fall behind easily. And since the spring module is optional and newly introduced I'd expect that dependency conflicts are rather unlikely for the foreseeable future 😉

Comment (beckermarc) [reply bom/pom.xml:89]: I think what @MatKuhr meant was that we should not add the Spring dependency to the BOM, but only to the newly introduced `sap-spring-properties` module. There we can of course reference a version directly.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java (hunk 3 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:76]: that's a good question!
I noticed that all property names at the bindings are lower-case. And the used libraries search for the bindings' properties in lower-case.
But Spring application properties names are usually in lower-camel-case.
I fell into this trap when externalizing these VCAP properties and adding them to my test `application.yaml` file. Then the bindings were not detected and/or the credentials could not be read.
I assumed a standard Spring Boot developer will do the same mistake. That's why I normalized the property keys.
Of course not the values...

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 4 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:36]: good Idea

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 5 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:41]: right

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java (hunk 3 above)
Comment (beckermarc) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:76]: > And the used libraries search for the bindings' properties in lower-case.

Can you point to that? Do they just happen to use lower-case property names or are they explicitly lowercasing their property keys before looking into the binding properties?

I am wondering if that is really the case for all libraries. In my opinion it should be possible to get the same property keys that you would get via VCAP_SERVICES, and a service might decide to use uppercase or camelcase property names. I wouldn't enforce any normalization, but keep the responsibility to the developer to add the property keys in the right and expected format.

File: bom/pom.xml (hunk 1 above)
Comment (drichtsteiger) [reply bom/pom.xml:89]: my bad, sorry

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java (hunk 3 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:76]: I look it up (that will take some time)

in general I completely agree to your opinion (and at places where these properties are looked up, it should be normalized to accept different 'styles' of property names)- but that was not the case in my tests...)

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 6 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:92]: I had the impression without tags it didn't work. I'll verify it again, as I'm not sure...

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 7 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:87]: again, I'll retest this too...

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java (hunk 3 above)
Comment (beckermarc) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:76]: Okay, but then that's a "bug"/limitation in these other libraries. A very generic service binding library like this shouldn't make such assumptions in my opinion.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java (hunk 11 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:26]: well, I could not think of any usage of this library outside CAP and this provides a clearer structure of the application properties.

anyway, what about "services" (like the environment variable "VCAP_SERVICES")

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java (hunk 3 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:76]: OK, let's remove it and I search for the places, where I think those libs should be improved.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java (hunk 11 above)
Comment (beckermarc) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:26]: `services` or `bindings` sounds good

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java (hunk 12 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:30]: That`s a good question I stumbled across too, when I wrote the unit tests. I just copied the code from CAP: com.sap.cds.framework.spring.config.runtime.SpringPropertiesProvider
I thought there must be a good reason for it that I don't know...

Comment (beckermarc) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:30]: In that class we also use `Bindable.ofInstance` and not only `Bindable.of`, therefore we need the additional `else if`. As you only use `Bindable.of` here, checking `bindResult.isBound()` is enough,

Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:30]: oh, details :-) thanks for the explanation

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 10 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:67]: Sounds reasonable. I took this over from `SapVcapServicesServiceBindingAccessor`. I guess there it is wrong as well then:
```
[...]
 .withNameKey("name")
 .withServiceName(serviceName)
[...]
```
`serviceName` is the JSON key = binding name from VCAP

Comment (beckermarc) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java]: In VCAP_SERVICES the JSON key is the service name (not the service instance or binding name). You also don't find a single object under that key, but an array containing multiple service instances/bindings.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    public void postProcessEnvironment( ConfigurableEnvironment environment, SpringApplication application )
+    {
+        Bindable<SapServiceBindingsProperties> bindable = Bindable.of(SapServiceBindingsProperties.class);
+
+        BindResult<?> bindResult = Binder.get(environment).bind("services", bindable);
+        if( bindResult.isBound() ) {
+            SapServiceBindingsPropertiesAccessor
+                .setServiceBindingsProperties(((SapServiceBindingsProperties) bindResult.get()).getServiceBindings());
+        } else {
+            log.info("Could not bind any service from the properties.");
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:31]:
```suggestion
            log.debug("Could not bind any service from the properties.");
```

This might happen if no `services` section is defined. I think we shouldn't log that information on every startup. Rather log positive things on `info` than negative things. If you'd like a log message on info rather move it to the case that properties are bound and inform about that.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Nonnull
+    @Override
+    public List<ServiceBinding> getServiceBindings()
+        throws ServiceBindingAccessException
+    {
+        List<ServiceBinding> serviceBindings = fetchServiceBindings();
+
+        serviceBindings
+            .forEach(binding -> logger.debug("Service binding {} found in properties.", binding.getServiceName().get()) //NOSONAR
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:38]:
```suggestion
            .forEach(binding -> logger.debug("Service binding {} found in properties.", binding.getName().get()) //NOSONAR
```

There might be multiple bindings for the same service with different names. Should this be logged every time service bindings are accessed?

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Nonnull
+    private List<ServiceBinding> fetchServiceBindings()
+    {
+        Map<String, ServiceBindingProperties> serviceBindingsProperties =
+            SapServiceBindingsPropertiesAccessor.getServiceBindingsProperties();
+
+        if( serviceBindingsProperties.isEmpty() ) {
+            logger.info("No service bindings found in properties.");
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:51]: Seems to be a similar log line as in the environment post processor if the result is not bound.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    private ServiceBinding toServiceBinding(
+        @Nonnull String serviceBindingName,
+        @Nonnull final ServiceBindingProperties serviceBindingProperties )
+    {
+        if( validateServiceBindingProperties(serviceBindingProperties) ) {
+            return DefaultServiceBinding
+                .builder()
+                .copy(Collections.emptyMap())
+                .withName(serviceBindingName)
+                .withServiceName(serviceBindingProperties.getName())
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:72]:
```suggestion
                .withServiceName(serviceBindingProperties.getServiceName())
```

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    @Nullable
+    private ServiceBinding toServiceBinding(
+        @Nonnull String serviceBindingName,
+        @Nonnull final ServiceBindingProperties serviceBindingProperties )
+    {
+        if( validateServiceBindingProperties(serviceBindingProperties) ) {
+            return DefaultServiceBinding
+                .builder()
+                .copy(Collections.emptyMap())
+                .withName(serviceBindingName)
```
Comment (beckermarc) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:71]: If `serviceBindingProperties.getName()` is set I think it should override using the properties key `serviceBindingName`.

File: bom/pom.xml
Comment (beckermarc) [inline bom/pom.xml:1]: Please restore file to state on `main`. The diff contains only unnecessary formatting changes

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 17 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:66]: yes, but why should one duplicate that information (property-name and then again at a the property "name")? I just removed the unnecessary property "name".

Comment (beckermarc) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:66]: We typically use that pattern in CAP for cases where the desired name can't be easily expressed as a YAML key. For example Spring Boot doesn't like `_` in YAML keys and you need to quote that with an ugly `'[MY_KEY_WITH_UNDERSCORES]'`. In general there are less restrictions on YAML keys than on YAML values. Therefore I'd suggest to keep it, but default to the YAML key in case the name is not explicitly given. Then it is a totally optional property, but provides flexibility in certain edge case scenarios.

**Comment by drichtsteiger** [comment]
who is going to update this project to Java 17? The pipeline must be adjusted and I do not have enough knowledge to do it properly.

**Comment by MatKuhr** [comment]
> who is going to update this project to Java 17? The pipeline must be adjusted and I do not have enough knowledge to do it properly.

I raised #244 for this and it looks rather straightforward, so hopefully this PR can be unblocked soon 😉

File: sap-spring-properties/pom.xml [hunk 18]
```diff
@@ -0,0 +1,88 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>com.sap.cloud.environment.servicebinding</groupId>
+        <artifactId>java-parent</artifactId>
+        <version>0.10.5</version>
+    </parent>
+    <artifactId>java-sap-spring-properties</artifactId>
```
Comment (newtork) [inline sap-spring-properties/pom.xml:9]: (Major)

Please correct me if I'm wrong, but the term "property" (or "properties") only reflect on the one file format. I think Spring call it "(Application) Configuration" and `.properties` is just one file format. Whereas `.yaml` would be even more prominent, right?

If not, then we should challenge the artifact id, since the name cannot be changed after release.

Comment (drichtsteiger) [reply sap-spring-properties/pom.xml:9]: well, it's the official term of the thing: https://docs.spring.io/spring-boot/how-to/properties-and-configuration.html
If people really struggle with this then they will run into much more severe problems not related to this lib at all...

File: sap-spring-properties/pom.xml [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        </dependency>
+    </dependencies>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>16</source>
+                    <target>16</target>
```
[the remark this answers is not in the document]
Comment (drichtsteiger) [reply sap-spring-properties/pom.xml:82]: and reverted then...

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+         * tags of the service binding
+         */
+        private String[] tags = new String[0];
+
+        /**
+         * properties for credentials
+         */
+        private Map<String, Object> credentials = new HashMap<>();
+
+        public String getName()
```
[the remark this answers is not in the document]
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:53]: I asked already if I can add Lombok to this project but got no answer so far. If it is fine I will do so.

[the remark this answers is not in the document]
Comment (newtork) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapServiceBindingsProperties.java:53]: I'm sorry, I assumed we already had _lombok_ in this repo.

Honestly, I would refrain from introducing _lombok_ as part of this PR.
For now we should accept the boilerplate until we have additional reason for _lombok_.

As alternative you could consider JDK17 `record` types :)
No priority though

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    private boolean validateServiceBindingProperties(
+        @Nonnull final ServiceBindingProperties serviceBindingProperties,
+        @Nonnull final String serviceBindingName )
+    {
+        boolean isValid = true;
+        if( serviceBindingProperties.getServiceName() == null || serviceBindingProperties.getServiceName().isEmpty() ) {
+            logger.error("Service binding properties of {} with no service name detected.", serviceBindingName);
+            isValid = false;
```
Comment (newtork) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:88]: (Minor/Nitpick)

You could've also `return` early.
But I would consider this _authors choice_. We don't have a rule for that.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+
+import com.sap.cloud.environment.servicebinding.SapServiceBindingsProperties;
+
+/* An {@link EnvironmentPostProcessor} that binds service binding properties to {@link SapServiceBindingsProperties}. */
+public class SapServiceBindingEnvironmentPostProcessor implements EnvironmentPostProcessor
+{
+    private static final Logger log = LoggerFactory.getLogger(SapServiceBindingEnvironmentPostProcessor.class);
+
+    @Override
+    public void postProcessEnvironment( ConfigurableEnvironment environment, SpringApplication application )
```
Comment (newtork) [inline sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java:20]: (Minor/Comment)

Strange, it looks like `@Nonnull final` is not being enforced here, but it should be.
I thought our pipeline would check for that, I'll investigate...

In the meantime, I'm sorry for sounding pedantic, but we try to declare variables and arguments with `final` wherever possible, as well as declaring them `@Nonnull` or `@Nullable`. The latter also applies to method signatures - for returned objects, excluding void and primitives.

These rules shall only be enforced on `public` API.

File: sap-spring-properties/src/test/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessorTest.java [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+                .filter(
+                    binding -> binding.getServiceName().isPresent()
+                        && binding.getServiceName().get().equals("service-manager"))
+                .collect(
+                    Collectors
+                        .toMap(serviceBinding -> serviceBinding.getName().get(), serviceBinding -> serviceBinding));
+        assertEquals(2, serviceManagerBindings.size());
+
+        ServiceBinding serviceManagerBinding1 = serviceManagerBindings.get("service-manager-test");
+        assertEquals("service-manager", serviceManagerBinding1.getServiceName().get());
```
Comment (newtork) [inline sap-spring-properties/src/test/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessorTest.java:126]: (Comment/Nitpick)

No action required. I just want to advertise `org.assertj.core.api.Assertions.assertThat` API as we use in other test classes. _AssertJ_ has powerful convenience methods that can be used for asserting. Great tool.

File: bom/pom.xml
Comment (newtork) [reply bom/pom.xml:1]: Before merging, I would be happy if you reverted the unexpected formatting changes.
The PR is in almost in perfect shape.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java (hunk 21 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessor.java:88]: I intentionally didn't. Validation failures indicate errors at the application's configuration and I would like to list all at once, in case there are more.

File: sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java (hunk 22 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/main/java/com/sap/cloud/environment/servicebinding/environment/SapServiceBindingEnvironmentPostProcessor.java]: good hint!

File: sap-spring-properties/src/test/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessorTest.java (hunk 23 above)
Comment (drichtsteiger) [reply sap-spring-properties/src/test/java/com/sap/cloud/environment/servicebinding/SapSpringPropertiesServiceBindingAccessorTest.java:126]: will have a look. thank you for the hint!

**Review by newtork — DISMISSED** [review]
LGTM

**Comment by newtork** [comment]
Please run `mvn formatter:format -pl :java-sap-spring-properties`

**Comment by drichtsteiger** [comment]
> ~Please run `mvn formatter:format -pl :java-sap-spring-properties`~

sorry, forgot that!

**Review by newtork — APPROVED** [review]
